### PR TITLE
Fix SpatioTemporalSparseCVI and add spatiotemporal tests

### LINKED ISSUE #22: Spatiotemporal variational demo model always emits mean 0
**Describe the bug**

With `SpatioTemporalSparseCVI`, after training, `space_time_predict_f` emits only zeros for the mean in the demo.

**To reproduce**

1. Run the complete [spatiotemporal variational example](https://github.com/secondmind-labs/markovflow/blob/develop/docs/notebooks/spatio_temporal.py)
2. Call the following:
```python
>>> mu_f, var_f = model.space_time_predict_f(X_grid)
>>> mu_f
<tf.Tensor: shape=(2500, 1), dtype=float64, numpy=
array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])>
>>> 
>>> np.max(np.abs(mu_f.numpy()))
0.0
```

**Expected behaviour**

It is surprising that the predicted mean is zero everywhere. This is visible in the plots in the notebook, where the prediction looks nothing like the data even after training: https://secondmind-labs.github.io/markovflow/notebooks/spatio_temporal.html?highlight=spatiotemporalsparsecvi:

I expected the mean to look like it had learned something from the data. 

**System information**
 - OS: Pop!_OS 22.04 LTS
 - Python version: 3.7
 - Markovflow version: develop branch
 - TensorFlow version: 2.4.3
 - GPflow version: 2.2.1

**Additional context**
If I modify the demo to instead use `SpatioTemporalSparseVariational`, `space_time_predict_f` does not predict all zeros. After fitting for many more iterations I get a reasonable-looking result:

It appears that 73232004e453baffeff871733c91b6a7fe7432ea may have been the commit that broke `SpatioTemporalSparseCVI` If we check out the previous commit (fc82c0abe777dd29edbb1eb7186a40579a83519b), this is the plot at the end of the notebook:

I suspect the culprit is these lines:
https://github.com/secondmind-labs/markovflow/blob/67e80c8117d256767c5ff255af29a5f5ecb37ee9/markovflow/models/spatio_temporal_variational.py#L433-L437 
Since `dist_q` is a property that is generated each time it is called, perhaps declaring `_posterior` in the constructor makes it out-of-sync with the trained parameters?

If we just convert these lines to
```python
    @property
    def posterior(self) -> PosteriorProcess:
        return ConditionalProcess(
            posterior_dist=self.dist_q,
            kernel=self.kernel,
            conditioning_time_points=self.inducing_time,
        )
```
then we again get a non-zero mean with predictions that look like the data:

### PR DESCRIPTION
**Pull request by sethaxen**
This PR fixes #22

As suggested by @vincentadam87 in https://github.com/secondmind-labs/markovflow/issues/22#issuecomment-1269856451, it also adds integration tests comparing the elbo of the variational model against the log-likelihood of `GPR`. It also checks that the predicted means are the same (predicted variances do not agree; should they?)

### CODE REVIEW

File: tests/integration/models/test_spatio_temporal_variational.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        kernel_space=gpflow.kernels.Matern32(lengthscales=1, variance=1, active_dims=[0]),
+        kernel_time=Matern12(lengthscale=1, variance=1),
+        likelihood=Gaussian(),
+    )
+    for t in (
+        params["kernel_space"].trainable_variables
+        + params["kernel_time"].trainable_variables
+        + params["likelihood"].trainable_variables
+    ):
+        t._trainable = False
```
Comment (vincentadam87) [inline tests/integration/models/test_spatio_temporal_variational.py]: I had in mind to only test the inference part in st model,
which is done by calling update_sites.
In this scenario, there is no need to use any external optimizer, so no need to fixed trainable to False

File: tests/integration/models/test_spatio_temporal_variational.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+def _setup_gpr_model_fixture(st_data):
+    kernel = gpflow.kernels.Product(
+        [
+            gpflow.kernels.Matern32(lengthscales=1, variance=1, active_dims=[0]),
+            gpflow.kernels.Matern12(lengthscales=1, variance=1, active_dims=[1]),
+        ]
+    )
+    model = GPR(data=st_data, kernel=kernel)
+    for t in model.trainable_variables:
+        t._trainable = False
```
Comment (vincentadam87) [inline tests/integration/models/test_spatio_temporal_variational.py]: Same for GPR.
GPR directly gives you the likelihood & posterior given all the hyperparameters
nothing to be learned in this test setting for the GPR model

File: tests/integration/models/test_spatio_temporal_variational.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+@pytest.fixture(name="st_model_params")
+def _setup_spatiotemporal_params_fixture(st_data):
+    X, _ = st_data
+    params = dict(
+        inducing_space=np.unique(X[:, 0]).reshape(-1, 1),
+        inducing_time=np.unique(X[:, 1]),
+        kernel_space=gpflow.kernels.Matern32(lengthscales=1, variance=1, active_dims=[0]),
+        kernel_time=Matern12(lengthscale=1, variance=1),
+        likelihood=Gaussian(),
```
Comment (vincentadam87) [inline tests/integration/models/test_spatio_temporal_variational.py:57]: maybe you want to set the learning_rate=1
currently it is set to 0.1 and I m not sure it converges in the 10 steps you have here.

**Review by vincentadam87 — COMMENTED** [review]
Looks good
If you could have a small docstrings description of what the test does, that would be great.
I m surprised about the non matching posterior variance, can you fix the optimization (higher lr or more iterations for cvi and let me know if the test still fails?)
Maybe focus on cvi for now, the other method is less stable so cvi should be easier to get to work

File: tests/integration/models/test_spatio_temporal_variational.py (hunk 3 above)
Comment (sethaxen) [reply tests/integration/models/test_spatio_temporal_variational.py:57]: The learning rate is set to 1 here: https://github.com/secondmind-labs/markovflow/blob/5229ac63e8e18856c59002a73404161f2e536c49/tests/integration/models/test_spatio_temporal_variational.py#L111

It seems to converge for ELBO, mean, and variance after a single step. Running even for 1,000 steps gives no different result than 1 step.

**Comment by sethaxen** [comment]
> I m surprised about the non matching posterior variance, can you fix the optimization (higher lr or more iterations for cvi and let me know if the test still fails?)

Here's the disagreement in variance (the same whether running `update_sites` once or 1,000 times):
```python
>       assert np.allclose(gpr_var, st_var, atol=atol, rtol=rtol)
E       assert False
E        +  where False = <function allclose at 0x7f1aa6240440>(<tf.Tensor: shape=(4, 1), dtype=float64, numpy=\narray([[0.45350882],\n       [0.45350882],\n       [0.45350882],\n       [0.45350882]])>, <tf.Tensor: shape=(4, 1), dtype=float64, numpy=\narray([[0.48707519],\n       [0.41994245],\n       [0.48707519],\n       [0.41994245]])>, atol=1e-06, rtol=1e-06)
```
So GPR predicts 4 identical marginal variances of `0.45350882`, while spatiotemporal CVI predicts 2 unique variances `[[0.48707519], [0.41994245]]`, each appearing twice. Interestingly, the average of these two values is `0.45350882`, exactly GPR's prediction.

Varying the learning rate from 0.01 to 1 and varying the number of `update_sites` calls from 10 to 1_000 still causes just the variance test to fail.

**Comment by vincentadam87** [comment]
Ok that looks good
I will take an hour to explore the problem of non-matching variances

**Review by vincentadam87 — APPROVED** [review]
